Cython/Utility/FunctionArguments.c

```diff
@@ -241,15 +241,43 @@ static int __Pyx_ParseOptionalKeywords(
     int kwds_is_tuple = CYTHON_METH_FASTCALL && likely(PyTuple_Check(kwds));
 
     while (1) {
+#if CYTHON_AVOID_BORROWED_REFS
+        Py_XDECREF(key); key=0;
+        Py_XDECREF(value); value=0;
+#endif
         if (kwds_is_tuple) {
-            if (pos >= PyTuple_GET_SIZE(kwds)) break;
+            Py_ssize_t size;
+#if CYTHON_ASSUME_SAFE_MACROS
+            size = PyTuple_GET_SIZE(kwds);
+#else
+            size = PyTuple_Size(kwds);
+            if (size < 0) goto bad;
+#endif
+            if (pos >= size) break;
+#if CYTHON_ASSUME_SAFE_MACROS && !CYTHON_AVOID_BORROWED_REFS
             key = PyTuple_GET_ITEM(kwds, pos);
+#elif !CYTHON_AVOID_BORROWED_REFS
+            key = PyTuple_GetItem(kwds, pos);
+#else
+            key = __Pyx_PySequence_ITEM(kwds, pos);
+#endif
+#if !CYTHON_ASSUME_SAFE_MACROS
+            if (!key) goto bad;
+#endif
             value = kwvalues[pos];
+#if !CYTHON_AVOID_BORROWED_REFS
+            Py_INCREF(value);
+#endif
             pos++;
         }
         else
         {
             if (!PyDict_Next(kwds, &pos, &key, &value)) break;
+#if !CYTHON_AVOID_BORROWED_REFS
```

Comment:
Isn't this guard the wrong way round? And, should we really keep the two borrowed references around for so long below? We might want to incref both any way, don't you think? At least if the fast pointer comparison below fails and we need to do actual comparisons. I don't think those are always safe while holding on to borrowed references.

Reply:
The guard was definitely the wrong way round. I've changed it so key and value are universally reference counted in this function to avoid having to think too much about what's safe.